README.md
### 应用程序使用自动轮换的凭据访问数据库 所需要使用的 lib 和 相应的代码示例


#### 使用场景

在应用程序的开发过程中，数据库是最常被用到的用于存储业务数据的基础服务。为了能够对数据库的访问权限进行有效的控制，应用程序只有在获得具有合适权限的账号和密码时，才能够创建数据库连接。为了降低账号和密码泄露的风险，应用程序需要定期更换使用的账号和密码信息。

SSM（凭据管理系统）能够完美的适用于上述场景，使得应用程序在避免频繁更换账号密码的麻烦的同时，保证业务数据的安全。



#### SSM SDK功能特性

基于 SSM（凭据管理系统）提供的数据库凭据功能，以及数据库账号安全管理的最佳实践，我们开发了一套可以方便应用程序集成的SSM SDK。应用程序只需要调用SSM SDK，传入必要的数据库连接相关的参数以及凭据名称，就可以直接获取到一个可用的数据库连接，而不必关心账号密码的获取和轮转的实现细节。

#### 风险提示

SSM在对数据库凭据进行周期性轮转的时候，会更新账号和密码。请严格按照**SDK使用说明**来使用SSM SDK。**不要在除了SSM SDK内部逻辑之外的任何地方缓存获取到的数据库连接，也不要缓存获取到的凭据中的任何信息**，以避免账号密码失效导致的数据库连接失败情况的发生。

### 支持的 Golang 版本
golang 1.13 及以上版本

#### 前置条件

1. 已在腾讯云平台开通了SSM服务（[开通SSM服务](https://console.cloud.tencent.com/ssm/cloud)）
2. 已在腾讯云平台购买了至少一台云数据库实例（目前只支持MySQL实例），完成了数据库的初始化，并创建了至少一个database。([MySQL控制台](https://console.cloud.tencent.com/cdb))
3. 已在SSM控制台创建了一个[数据库凭据](https://cloud.tencent.com/document/product/1140/57647)，并和指定的数据库做了关联。（具体操作指南参见：[创建数据库凭据](https://cloud.tencent.com/document/product/1140/57648)）
4. 已在腾讯云平台的[访问管理（CAM）控制台](https://console.cloud.tencent.com/cam/overview)创建了一个能够访问SSM凭据资源和MySQL实例资源的子账号，并给该子账号分配了[API密钥](https://console.cloud.tencent.com/cam/capi)（以便获取SecretId 和 SecretKey 用于API的访问）


#### SDK使用的代码示例
1. 获取go模块：

   （1）运行命令：go get github.com/tencentcloud/ssm-rotation-sdk-golang/lib

   （2）在golang项目中引入：		

   ```golang
   import (
      "github.com/tencentcloud/ssm-rotation-sdk-golang/lib/db"
      "github.com/tencentcloud/ssm-rotation-sdk-golang/lib/ssm"
   )
   ```

2. 初始化 DynamicSecretRotationDb 对象

   ```golang
   dbConn = &db.DynamicSecretRotationDb{}
   err := dbConn.Init(&db.Config{
      DbConfig: &db.DbConfig{
         MaxOpenConns:        100,
         MaxIdleConns:        50,
         IdleTimeoutSeconds:  100,
         ReadTimeoutSeconds:  5,
         WriteTimeoutSeconds: 5,
         SecretName:          "test",          // 凭据名
         IpAddress:           "127.0.0.1",     // 数据库实例的访问地址
         Port:                58366,           // 数据库实例的访问端口
         DbName:              "database_name", // 指定具体的数据库名，如果为空，则只连接到数据库实例，不连接具体的数据库
         ParamStr:            "charset=utf8&loc=Local", // 数据库连接相关的配置参数
      },
      SsmServiceConfig: &ssm.SsmAccount{
         SecretId:  "SecretId",     // 子账号的SecretId
         SecretKey: "SecretKey",    // 子账号的SecretKey
         Region:    "ap-guangzhou", // 选择凭据实际所存储的地域，请根据实际情况填写，示例中给出的是广州地区。
      },
      WatchChangeInterval: time.Second * 10, // 监控凭据内容发生变化的间隔，需小于凭据轮转周期，一般间隔时间设置为10秒~60秒之间为宜
   })
   ```

3. 获取数据库连接

   ```golang
   c := dbConn.GetConn() // 警告：每次需要访问数据库时，都需要调用GetConn()来获取最新的DB连接，请不要在业务代码中缓存此对象，以免DB访问失败！
   if err := c.Ping(); err != nil { // 示例中只是调用ping()方法测试账号密码的可用性。实际业务中，这里就可以执行具体的db操作了。
      log.Fatal("failed to access db with err: ", err)
      return
   }
   ```

4. 关闭数据库连接

   ```golang
   c.Close() // 当应用程序退出时，可主动关闭数据库连接。这是个通用的操作，和数据库凭据没有直接关系
   ```




CHANGELOG.md
# Changelog

本项目的所有重要变更将记录在此文件中。

格式基于 [Keep a Changelog](https://keepachangelog.com/zh-CN/1.0.0/)，
本项目遵循 [语义化版本](https://semver.org/lang/zh-CN/)。

## [1.0.0] - 2026-03-22

### 新增

- 支持从 SSM 自动获取数据库凭据
- 定期监控凭据变化，自动更新数据库连接
- 线程安全的连接池管理（基于 Go database/sql）
- 三种认证方式：CamRole（推荐）、Temporary、Permanent
- 健康检查 API（`IsHealthy()` / `GetHealthCheckResult()`）
- 轮转时旧连接延迟退休机制（RotationGracePeriod）
- Watch 启动随机初始延时，避免多实例请求风暴
- 支持自定义 SSM 接入点（endpoint）
- 支持自定义读写超时参数
- 支持额外数据库连接参数透传


lib/ssm/requester.go
package ssm

import (
	"encoding/json"
	"fmt"
	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common"
	"github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/common/profile"
	ssm "github.com/tencentcloud/tencentcloud-sdk-go/tencentcloud/ssm/v20190923"
	"log"
	"time"
)

type DbAccount struct {
	UserName string `json:"UserName"`
	Password string `json:"Password"`
}

func GetCurrentAccount(secretName string, ssmAcc *SsmAccount) (*DbAccount, error) {
	secretValue, err := getCurrentProductSecretValue(secretName, ssmAcc) // secretValue里面存储了用户名和密码，格式为：userName:password
	if err != nil {
		log.Println("failed to GetCurrentProductSecretValue, err= ", err)
		return nil, err
	}
	// secretValue 是 JSON格式的字符串，形如： {"UserName":"test_user","Password":"test_pwd"}
	log.Println("secret value: ", secretValue)
	if len(secretValue) == 0 {
		return nil, fmt.Errorf("no valid account info found because secret value is empty")
	}
	account := &DbAccount{}
	if err := json.Unmarshal([]byte(secretValue), account); err != nil {
		log.Println("err when parse secretValue in json format, err= ", err)
		return nil, fmt.Errorf("invalid secret value format")
	}
	return account, nil
}

func getCurrentProductSecretValue(secretName string, ssmAcc *SsmAccount) (string, error) {
	log.Printf("get value for secretName=%v", secretName)
	startTime := time.Now()
	client, err := getClient(ssmAcc.SecretId, ssmAcc.SecretKey, ssmAcc.Url, ssmAcc.Region)
	if err != nil {
		log.Fatal(" create ssm client error: ", err)
		return "", fmt.Errorf("create ssm HTTP client error: %s", err)
	}
	request := ssm.NewGetSecretValueRequest()
	request.SecretName = &secretName
	request.VersionId = common.StringPtr("SSM_Current")
	log.Printf("GetSecretValue request=%s", request.ToJsonString())
	rsp, err := client.GetSecretValue(request)
	if err != nil {
		log.Fatal(" ssm GetSecretValue error: ", err.Error())
		return "", fmt.Errorf("ssm GetSecretValue error: %s", err.Error())
	}
	if rsp.Response.SecretString == nil {
		log.Println("Secret Value is nil")
		return "", nil
	}
	log.Printf("GetCurrentProductSecretValue cost time: %d", time.Since(startTime))
	return *rsp.Response.SecretString, nil
}

func getClient(secretId string, secretKey string, url string, region string) (*ssm.Client, error) {
	credential := common.NewCredential(secretId, secretKey)
	httpProfile := profile.NewHttpProfile()
	httpProfile.ReqMethod = "POST"
	if len(url) != 0 {
		httpProfile.Endpoint = url
	}
	cpf := profile.NewClientProfile()
	cpf.HttpProfile = httpProfile
	client, err := ssm.NewClient(credential, region, cpf)
	return client, err
}

type SsmAccount struct {
	SecretId  string `yaml:"secretId"`
	SecretKey string `yaml:"secretKey"`
	Url       string `yaml:"url"`
	Region    string `yaml:"region"`
}


lib/db/dynamicSecretRotationDbConn.go
package db

import (
	"database/sql"
	"fmt"
	"github.com/tencentcloud/ssm-rotation-sdk-golang/lib/ssm"
	"log"

	"sync/atomic"
	"time"
)

type DynamicSecretRotationDb struct {
	config *Config      // 初始化配置
	dbConn atomic.Value // 存储的是 ConnCache 结构体
}

type ConnCache struct {
	ConnStr string // 缓存的当前正在使用的连接信息
	Conn    *sql.DB
}

type Config struct {
	DbConfig            *DbConfig
	SsmServiceConfig    *ssm.SsmAccount
	WatchChangeInterval time.Duration
}

type DbConfig struct {
	MaxOpenConns        int
	MaxIdleConns        int
	IdleTimeoutSeconds  int
	ReadTimeoutSeconds  int
	WriteTimeoutSeconds int
	SecretName          string
	IpAddress           string
	Port                uint64
	DbName              string
	ParamStr            string // 例如：charset=utf8&loc=Local
}

func (c *Config) BuildConnStr() (string, error) {
	account, err := ssm.GetCurrentAccount(c.DbConfig.SecretName, c.SsmServiceConfig) // secretValue里面存储了用户名和密码，格式为：userName:password
	if err != nil {
		return "", err
	}
	// connection string 的格式： {user}:{password}@tcp({ip}:{port})/{dbName}?charset=utf8&loc=Local
	connStr := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s", account.UserName, account.Password, c.DbConfig.IpAddress, c.DbConfig.Port, c.DbConfig.DbName)
	if len(c.DbConfig.ParamStr) != 0 {
		connStr = fmt.Sprintf("%s?%s", connStr, c.DbConfig.ParamStr)
	}
	return connStr, nil
}

/**
在服务初始化的时候，可调用本方法来完成数据库连接的初始化。
本方法根据提供的凭据相关的信息（服务账号，凭据名），获得真实的数据库用户名和密码信息，然后生成数据库连接
*/
func (d *DynamicSecretRotationDb) Init(config *Config) error {
	d.config = config
	err := d.initDbConn()
	if err != nil {
		return err
	}
	log.Println("succeed to init dbConn: ", d.GetConn())
	go d.watchSecretChange()
	return nil
}

/**
调用方每次访问db时，需通过调用本方法获取db连接。
注意：请不要在调用端缓存获取到的 *sql.DB, 以便确保在凭据发生轮换后，能及时的获得到最新的用户名和密码，防止由于用户名密码过期，而造成数据库连接失败！
*/
func (d *DynamicSecretRotationDb) GetConn() *sql.DB {
	if conn, ok := d.dbConn.Load().(*ConnCache); ok {
		log.Println("GetConn, connStr=", conn.ConnStr)
		return conn.Conn
	}
	return nil
}

func (d *DynamicSecretRotationDb) getCurrentConnStr() string {
	conn := d.dbConn.Load().(*ConnCache)
	log.Println("GetConn, connStr=", conn.ConnStr)
	return conn.ConnStr
}

func (d *DynamicSecretRotationDb) initDbConn() error {
	var err error
	connStr, err := d.config.BuildConnStr()
	if err != nil {
		return err
	}
	tmpDbConn, err := sql.Open("mysql", connStr)
	if err != nil {
		return fmt.Errorf("connect to cdb error: %s", err)
	}
	dbConfig := d.config.DbConfig
	tmpDbConn.SetMaxOpenConns(dbConfig.MaxOpenConns)
	tmpDbConn.SetMaxIdleConns(dbConfig.MaxIdleConns)

	connLifeTime := time.Duration(dbConfig.IdleTimeoutSeconds) * time.Second
	tmpDbConn.SetConnMaxLifetime(connLifeTime)

	err = tmpDbConn.Ping()
	if err != nil {
		return fmt.Errorf("ping cdb error: %s ", err)
	}
	// 将有效的 connStr 缓存起来
	curConn := d.GetConn()
	d.dbConn.Store(&ConnCache{
		ConnStr: connStr,
		Conn:    tmpDbConn,
	})
	if curConn != nil {
		if err := curConn.Close(); err != nil {
			log.Println("failed to close connection, err= ", err)
			return err
		}
	}
	return nil
}

func (d *DynamicSecretRotationDb) watchSecretChange() {
	t := time.Tick(d.config.WatchChangeInterval)
	for {
		select {
		case <-t:
			d.watchChange()
		}
	}
}

func (d *DynamicSecretRotationDb) watchChange() {
	connStr, err := d.config.BuildConnStr()
	if err != nil {
		log.Println("failed to build connStr, err= ", err)
		return
	}
	if connStr == d.getCurrentConnStr() {
		log.Println("db connStr is not changed")
		return
	}

	log.Printf("connstr is changing from [%s] to [%s] \n", d.getCurrentConnStr(), connStr)
	err = d.initDbConn()
	if err != nil {
		log.Println("failed to initDbConn, err=", err)
		return
	}
	log.Printf("**** succeed to change dbConn, new connStr=%s  **** \n", d.getCurrentConnStr())

}


demo/demo.go
package main

import (
	"log"

	_ "github.com/go-sql-driver/mysql"
	"github.com/tencentcloud/ssm-rotation-sdk-golang/lib/db"
	"github.com/tencentcloud/ssm-rotation-sdk-golang/lib/ssm"

	"time"
)

var dbConn *db.DynamicSecretRotationDb

func main() {
	// 初始化数据库连接
	dbConn = &db.DynamicSecretRotationDb{}
	err := dbConn.Init(&db.Config{
		DbConfig: &db.DbConfig{
			MaxOpenConns:        100,
			MaxIdleConns:        50,
			IdleTimeoutSeconds:  100,
			ReadTimeoutSeconds:  5,
			WriteTimeoutSeconds: 5,
			SecretName:          "test",          // 凭据名
			IpAddress:           "127.0.0.1",     // 数据库地址
			Port:                58366,           // 数据库端口
			DbName:              "database_name", // 可以为空，或指定具体的数据库名
			ParamStr:            "charset=utf8&loc=Local",
		},
		SsmServiceConfig: &ssm.SsmAccount{
			SecretId:  "SecretId",     // 需填写实际可用的SecretId
			SecretKey: "SecretKey",    // 需填写实际可用的SecretKey
			Region:    "ap-guangzhou", // 选择凭据所存储的地域
		},
		WatchChangeInterval: time.Second * 10, // 多长时间检查一下 凭据是否发生了轮转
	})
	if err != nil {
		log.Fatal("failed to init dbConn, err=", err)
		return
	}
	// 模拟业务处理中，每过一段时间（一般是几毫秒），需要拿到db连接，来操作数据库的场景
	t := time.Tick(time.Second)
	for {
		select {
		case <-t:
			accessDb()
		}
	}
}

func accessDb() {
	log.Println("--- accessDb start")
	c := dbConn.GetConn()
	if err := c.Ping(); err != nil {
		log.Fatal("failed to access db with err: ", err)
		return
	}
	log.Println("--- succeed to access db")
}
